Updates to the .NET Client Guide (#1401)
RedisLabs/redislabs-docs: content/rs/references/client_references/client_csharp.md

---
Title: Using Redis with .NET
description:
weight:
alwaysopen: false
categories: ["RS"]
---

> In order to use Redis in .NET you need a .NET Redis client. In following sections, we demonstrate the use of [StackExchange.Redis](https://github.com/StackExchange/StackExchange.Redis), a general purpose Redis client. Additional .NET clients for Redis can be found under the [C# section](https://redis.io/clients#c-sharp) of the Redis Clients page.

Reviewer: ```suggestion
In order to use Redis in .NET, you need a .NET Redis client. This article shows how to use [StackExchange.Redis](https://github.com/StackExchange/StackExchange.Redis), a general purpose Redis client.  More .NET Redis clients can be found in the [C# section](https://redis.io/clients#c-sharp) of the Redis Clients page.
```

## Installing StackExchange.Redis

There are several ways to install this package including:

> 1. With the .NET CLI: 

Reviewer: Add link to .NET CLI intro article.

```sh
dotnet add package StackExchange.Redis
```

> 2. with the package manager console: 

Reviewer: Add link to package manager intro article.

```sh
`PM> Install-Package StackExchange.Redis`
```

> 3. With the NuGet GUI in Visual Studio   

Reviewer: Add link to NuGet client install instructions)

## Opening a Connection to Redis Using StackExchange.Redis

> The following code creates a connection to Redis using StackExchange.Redis in the context of a trivial console application:

Reviewer: ```suggestion
The following code creates a connection to Redis using StackExchange.Redis in the context of a console application:
```

```csharp
using StackExchange.Redis;
using System;
using System.Threading.Tasks;

namespace ReferenceConsoleRedisApp
{
    class Program
    {
        static readonly ConnectionMultiplexer redis = ConnectionMultiplexer.Connect(
            new ConfigurationOptions{
                EndPoints = {"localhost:6379"}                
            });
        static async Task Main(string[] args)
        {
            var db = redis.GetDatabase();
            var pong = await db.PingAsync();
            Console.WriteLine(pong);
        }
    }
}

```

The above example assumes that you have a Redis Server running locally.

> To configure the connection to your environment, adjust the parameters in the [ConfigurationOptions](https://stackexchange.github.io/StackExchange.Redis/Configuration) object appropriately. For the remainder of the examples the configuration will simply be `localhost`.

Reviewer: ```suggestion
To configure the connection to your environment, adjust the parameters in the [ConfigurationOptions](https://stackexchange.github.io/StackExchange.Redis/Configuration) object appropriately. For the remainder of the examples, the configuration will simply be `localhost`.
```

## Connection Pooling with StackExchange.Redis

> While StackExchange.Redis does not provide direct means for conventional connection pooling, we recommend you **share and reuse** the ConnectionMultiplexer object. The ConnectionMultiplexer object should not be created per operation - it is to be created only once at the beginning and reused for the duration of the run. ConnectionMultiplexer is thread-safe so it can be safely shared between threads. For more information, refer to StackExchange.Redis’s [Basic Usage document](https://stackexchange.github.io/StackExchange.Redis/Basics).

Reviewer: ```suggestion
StackExchange.Redis does not support conventional connection pooling.  We recommend **sharing and reusing** the ConnectionMultiplexer object. The ConnectionMultiplexer object should not be created for each operation.  Instead, create an instance at the beginning and then reuse the object throughout your process. ConnectionMultiplexer is thread-safe so it can be safely shared between threads. For more information, see the  [Basics document](https://stackexchange.github.io/StackExchange.Redis/Basics).
```

Reviewer: Why is sharing and reusing bolded here?  It shouldn't be unless there's a semantic reason.  (It's not a term of art, so it doesn't need to be defined as one, nor does it refer to a UX element.)

Author: It was in there originally, I think I meant to underline the fact that that's really the behavior you should be using - I could  go either way on having it bolded - if the style guides specifically call for it not to be bolded that's a good enough reason to un-bold it.

Reviewer: We use bolding and emphasis for specific situations.  I'd remove unless there's a good reason not to.

Author: done

## Dependency Injection of the ConnectionMultiplexer

> As the `ConnectionMultiplexer` must be shared and reused within a runtime, it's recommended that you use [dependency injection](https://docs.microsoft.com/en-us/dotnet/core/extensions/dependency-injection) to pass it where it's needed. There's a few flavors of dependency injection depending on what you're using.

Reviewer: When referring to specific classes, methods, etc, try to link to a doc explaining it on first use  Example: `[ConnectionManager](https://stackexchange.github.io/StackExchange.Redis/Basics)`.  This should be as close to a reference page as possible.  (Here, we have to link to the basics page, as there doesn't appear to be a reference page for the ConnectionMultiplexer class.)

### ASP.NET Core 

> A single `ConnectionMultiplexer` instance should be shared throughout the runtime, therefore the most appropriate pattern to use for dependency injection in .NET Core is to inject it as a singleton. In `Startup.cs`, add the connection logic to your `ConfigureServices` method, e.g.

Reviewer: ```suggestion
A single `ConnectionMultiplexer` instance should be shared.  Use the singleton pattern to to inject your instance as a dependency in .NET Core. In `Startup.cs`, add the connection logic to your `ConfigureServices` method:
```

Author: I disagree with this - "Singleton Pattern" refers to a very specific design pattern in .NET where you have a Static instance object - here we are referring to using ASP.NET Core's Singleton service injection functionality. How about: 

```
A single `ConnectionMultiplexer` instance should be shared throughout the runtime. Use the `AddSingleton` method of `IServiceCollection` to inject your instance as a dependency in .NET Core when configuring your app's services in `Startup.cs`:
```

Reviewer: Looks good to me.

```csharp
public void ConfigureServices(IServiceCollection services)
{
    //Configure other services up here
    var multiplexer = ConnectionMultiplexer.Connect("localhost");
    services.AddSingleton<IConnectionMultiplexer>(multiplexer);
}
```

> With the service registered, you can then inject the service into anything that allows dependency injection e.g. MVC Controllers, API Controllers, Blazor Server Components etc. . . By simply handing it to the constructor of that type, for example, if we wanted to pass it to a `RedisController` we would do so with the following:

Reviewer: ```suggestion
Once the service is registered, you can inject it into anything that allows dependency injection, such as MVC Controllers, API Controllers, Blazor Server Components, and more.  Simply hand the service to the appropriate controller.  Here's how to pass it to a `RedisController` instance:
```

```csharp
[Route("api/[controller]")]
[ApiController]
public class RedisController : ControllerBase
{
    private readonly IConnectionMultiplexer _redis;
    
    public RedisController(IConnectionMultiplexer redis)
    {
        _redis = redis;
    }

    [HttpGet("foo")]
    public async Task<IActionResult> Foo()
    {
        var db = _redis.GetDatabase();
        var foo = await db.StringGetAsync("foo");
        return Ok(foo.ToString());
    }
}
```

### Azure Functions

There are two types of Azure functions to consider [in-process](https://docs.microsoft.com/en-us/azure/azure-functions/functions-dotnet-class-library) and [out-of-process](https://docs.microsoft.com/en-us/azure/azure-functions/dotnet-isolated-process-guide). Both handle dependency injection differently.

#### In-Process Azure Functions

> In-process Azure Functions handle dependency injection very similarly to how it's handled in ASP.NET Core. Create a Startup.cs file, and have the class extend `FunctionStartup`, override the `Configure` method, an add the multiplexer as a Singleton service for the function:

Reviewer: ```suggestion
In-process Azure Functions handle dependency injection similarly to ASP.NET Core. You:
1. Create a `Startup.cs` file
2. Extend `FunctionStartup`
3. Override the `Configure` method
4. Add the multiplexer as a Singleton service for the function.

```

```csharp
using Microsoft.Azure.Functions.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection;
using StackExchange.Redis;
[assembly: FunctionsStartup(typeof(MyNamespace.Startup))]

namespace MyNamespace
{
    public class Startup : FunctionsStartup
    {
        public override void Configure(IFunctionsHostBuilder builder)
        {
            var muxer = ConnectionMultiplexer.Connect("localhost");
            builder.Services.AddSingleton<IConnectionMultiplexer>(muxer);            
        }
    }
}
```

#### Out-Of-Process Azure Functions

> Out-of-process Functions are newer, unlike in-process functions, dependency injection is handled in the `Main` method of `Program` class, modify the `HostBuilder` build pipeline to add a `ConfigureServices` call, adding a properly configured `ConnectionMultiplexer`. That method should look like: 

Reviewer: ```suggestion
Unlike in-process functions, out-of-process functions handle dependency injection in the `Main` method of `Program` class.  Modify the `HostBuilder` build pipeline to call `ConfigureServices` and add a properly to configure `ConnectionMultiplexer`. Here's an example: 
```

```csharp
public static void Main()
{
    var host = new HostBuilder()
        .ConfigureFunctionsWorkerDefaults()
        .ConfigureServices(s=>s.AddSingleton<IConnectionMultiplexer>(ConnectionMultiplexer.Connect("localhost"))) // add this line
        .Build();

    host.Run();
}
```

> In both instances you can pull out the Multiplexer by injecting it into the constructor of the class housing your functions:

Reviewer: ```suggestion
In both instances, you can pull out the Multiplexer by injecting it into the constructor of the class housing your functions:
```

```csharp
public class RedisTrigger
{
    private readonly IConnectionMultiplexer _redis;

    public RedisTrigger(IConnectionMultiplexer redis){
        _redis = redis;
    }
    [FunctionName("RedisTrigger")]
    public async Task<IActionResult> Foo(
        [HttpTrigger(AuthorizationLevel.Function, "get", "post", Route = null)] HttpRequest req,
        ILogger log)
    {
        var db = _redis.GetDatabase();
        var bar = await db.StringGetAsync("foo");

        return new OkObjectResult(bar.ToString());
    }
}
```

## Using SSL and StackExchange.Redis

> StackExchange.Redis is the first Redis client that natively supported TLS. The following code has the StackExchange Redis Client operate on a TLS connection:

Reviewer: ```suggestion
StackExchange.Redis was the first Redis client to natively support TLS, as shown here:
```

```csharp
using StackExchange.Redis;
using System;
using System.Security.Cryptography.X509Certificates;
using System.Threading.Tasks;

namespace ConnectToRedisWithTls
{
    class Program
    {

        const string PATH_TO_CERT_FILE = "c:\\PATH\\TO\\CERT.pfx";        

        static async Task Main(string[] args)
        {
            var configurationOptions = new ConfigurationOptions
            {
                EndPoints = { "localhost:6379" },
                Ssl = true
            };
            configurationOptions.CertificateSelection += delegate {
                var cert = new X509Certificate2(PATH_TO_CERT_FILE, "");
                return cert;
            };
            var redis = await ConnectionMultiplexer.ConnectAsync(configurationOptions);
            Console.WriteLine(redis.GetDatabase().Ping());
        }
    }
}
```
- Modify the `PATH_TO_CERT_FILE` to match the path to your certificate
- Modify the `Endpoints` setting to point at your endpoint(s)
- If necessary, add a `Password` to the Configuration options

> For information about how to run Redis Server with TLS enabled See the [TLS Support](https://redis.io/topics/encryption) section of the Redis documentation.

Reviewer: ```suggestion
To learn how to run Redis Server with TLS enabled, see [TLS Support](https://redis.io/topics/encryption).
```
